### PR TITLE
refactor(components): migrate Header component to TypeScript

### PR DESCRIPTION
**Pull request by hleote**
Addresses #516

## Purpose

Refactor Header component to TS - another step into having all components in TS

## Approach and changes

- [x] Change files extensions to ts/tsx
- [x] Refactor Header component
- [x] Refactor Title component
- [x] Refactor tests

## Definition of done

* [ ] Development completed
* [ ] Reviewers assigned
* [ ] Unit and integration tests
* [ ] Meets minimum browser support
* [ ] Meets accessibility requirements

### CODE REVIEW

**Review by connor-baer — APPROVED** [review]
This is fantastic, thanks @hleote! 🙌

[Eslint/Prettier doesn't seem to be happy yet](https://github.com/sumup-oss/circuit-ui/pull/806/checks?check_run_id=1936240466). Can you run `yarn lint` locally to see why that is, please? 

I've updated the PR title to follow the conventional commit format. When you squash-merge the PR, the PR title will be used as the commit message and then picked up by semantic-release for the changelog 🙂

**Comment by connor-baer** [comment]
FYI, the Header component will likely have breaking changes in the next major release, see https://github.com/sumup-oss/circuit-ui/issues/793.

**Comment by hleote** [comment]
> This is fantastic, thanks @hleote! 🙌
> 
> [Eslint/Prettier doesn't seem to be happy yet](https://github.com/sumup-oss/circuit-ui/pull/806/checks?check_run_id=1936240466). Can you run `yarn lint` locally to see why that is, please?
> 
> I've updated the PR title to follow the conventional commit format. When you squash-merge the PR, the PR title will be used as the commit message and then picked up by semantic-release for the changelog 🙂

By the way @JonathanMH also helped with this one 👍 💯

File: src/components/Header/Header.stories.tsx [hunk 1]
```diff
@@ -15,20 +15,14 @@
...
 export default {
   title: 'Components/Header',
   component: Header,
 };
 
-export const Base = (args) => (
-  <Header {...args}>
-    <Hamburger size="kilo" />
-  </Header>
-);
```
Comment (connor-baer) [inline src/components/Header/Header.stories.tsx:31]: What's your intention behind removing the hamburger here? 

The stories also serve as documentation and I believe it's important to document the use case with the hamburger.

Comment (hleote) [reply src/components/Header/Header.stories.tsx:31]: I will revert 🙂

**Review by connor-baer — APPROVED** [review]

**Comment by hleote** [comment]
@connor-baer what should we do about the codecov? Should I create some more tests? 🙂

**Comment by connor-baer** [comment]
The drop in coverage seems to come from [these lines](https://app.codecov.io/gh/sumup-oss/circuit-ui/compare/806/changes): 

Screen Shot 2021-02-23 at 12 08 41

There is [a test](https://github.com/sumup-oss/circuit-ui/pull/806/files#diff-eee5a49bbb6d5acd3a4afca78cf1c6c89b7a229de8bcb3ea290799ce1ae40448R46-R50) for it, so I'd ignore the check failure. Feel free to merge anyway.

**Pull request merged**